src/basic/process-util.h

```
@@ -142,15 +142,16 @@ void reset_cached_pid(void);
 int must_be_root(void);
 
 typedef enum ForkFlags {
-        FORK_RESET_SIGNALS = 1 << 0,
-        FORK_CLOSE_ALL_FDS = 1 << 1,
-        FORK_DEATHSIG      = 1 << 2,
-        FORK_NULL_STDIO    = 1 << 3,
-        FORK_REOPEN_LOG    = 1 << 4,
-        FORK_LOG           = 1 << 5,
-        FORK_WAIT          = 1 << 6,
-        FORK_NEW_MOUNTNS   = 1 << 7,
-        FORK_MOUNTNS_SLAVE = 1 << 8,
+        FORK_RESET_SIGNALS      = 1 << 0, /* Reset all signal handlers and signal mask */
+        FORK_CLOSE_ALL_FDS      = 1 << 1, /* Close all open file descriptors in the child, except for 0,1,2 */
+        FORK_DEATHSIG           = 1 << 2, /* Set PR_DEATHSIG in the child */
+        FORK_NULL_STDIO         = 1 << 3, /* Connect 0,1,2 to /dev/null */
+        FORK_REOPEN_LOG         = 1 << 4, /* Reopen log connection */
+        FORK_LOG                = 1 << 5, /* Log above LOG_DEBUG log level about failures */
+        FORK_WAIT               = 1 << 6, /* Wait until child exited */
+        FORK_NEW_MOUNTNS        = 1 << 7, /* Run child in its own mount namespace */
+        FORK_MOUNTNS_SLAVE      = 1 << 8, /* Make child's mount namespace MS_SLAVE */
+        FORK_RLIMIT_NOFILE_SAFE = 1 << 9, /* Set RLIMIT_NOFILE soft limit to 1K for select() compat */
```

I was wondering whether this should be opt-out. We need to reset the rlimit for everything we invoke that is not our code, which turns out to be about half of safe_fork calls. But the effect is not symmetrical: if the rlimit is too high, a program that does not expect this might silently fail. In the other direction, if the limit is lowered, there is almost no effect, because our programs that would make use of this usually bump the limit themselves. And if the limit was high, and a lot of fds were open, most of them should be cloexec. In the end, I think this should be opt out, since forgetting to do this is the most problematic.

---

We use safe_fork() at quite a few places where we don't fork() in the child, but just do some work out-of-process (because we want some isolation, join some namespace, deal with blocking syscalls, change identities, and so on). Hence, for us fork() really doesn't imply execve() right after. Moreover, in many cases where we do execve() after fork() we do fd manipulations first, and the RLIMIT_NOFILE readjustment should be done right after i figure. For both reasons together hence I opted to leave this opt-in rather make it opt-out.

THinking about this, maybe looking at this from the fork() perspective is kinda where the problem is and we should do things the other way round though: introduce safe_execve()  and make it opt-out there. Because after all the resetting of RLIMIT_NOFILE is kinda a pre-requirement of execve more than an a post-requirement of fork()